File: cal_amp_eps.m
function [iterEps, iterDel]= cal_amp_eps(numIter, whichComp, totEps, totDel, sampRate)

% numIter
% whichComp: 0 (linear), 1 (adv), 2 (zCDP)
% totEps
% totDel
% sampRate: minibatchsize/totalDatasize

%% 
x0 = 0.1;

if whichComp==0 % linear composition 
   
    fun = @(x) (totEps - numIter*log(1 + sampRate*(exp(x)-1)))^2; 
    iterEps = fmincon(fun,x0,[], [] ,[],[],1e-6,[]);
    iterDel = totDel/numIter;
%     eps_amp = np.log(1 + nu*(np.exp(x)-1))
elseif whichComp == 1 % Advanced composition 
    iterDel = 0.0001;
    delta = 0.0001; 
%     - numIter*sampRate*iterDel;
    if delta<0
        display('delta is less than 0')
    end
%     eps_amp = log(1 + sampRate*(exp(x)-1))
    fun = @(x) (totEps - sqrt(2*numIter*log(1/delta))*log(1 + sampRate*(exp(x)-1)) - numIter*log(1 + sampRate*(exp(x)-1))*(exp(log(1 + sampRate*(exp(x)-1)))-1))^2;
    iterEps = fmincon(fun,x0,[], [] ,[],[],1e-6,[]);

else % whichComp==2,  zCDP composition 
    iterDel = 0.0001;
    delta_amp = sampRate*iterDel;
    c2 = 2*log(1.25/delta_amp);
    delta = 0.0001;
%     rho = numIter*(eps_amp^2)/(2*c2)

    fun = @(x)(totEps - (numIter*((log(1 + sampRate*(exp(x)-1)))^2)/(2*c2) + 2*sqrt(numIter*((log(1 + sampRate*(exp(x)-1)))^2)/(2*c2)*log(1/delta))))^2; 
    iterEps = fmincon(fun,x0,[], [] ,[],[],1e-6,[]);
end

File: PPVI_BLR/cal_amp_eps.py
# __author__ = 'mijung'
# to calculate amplified epsilon according to which composition theorem to use
# wrote on the 31st of Aug, 2016

import matplotlib.pyplot as plt
import numpy as np
# import theano
# import theano.tensor as T
import scipy
from scipy.optimize import minimize_scalar

""" this is for (epsilon, delta)-DP mechanisms """


def f(x, total_eps, which_comp, J, nu):
	# calculate eps_amp first
	# trm1 = 1 + nu*(np.exp(x)-1)
	# trm2 = 1 - nu*(1-np.exp(-x))
	# # print trm1, trm2
	# eps_amp = np.log(trm1) - np.log(trm2)
	eps_amp = np.log(1 + nu * (np.exp(x) - 1))

	# (1) linear composition
	if which_comp == 0:
		# linear composition
		return (total_eps - J * eps_amp) ** 2

	elif which_comp == 1:
		# advanced composition
		delta_iter = 0.000001
		delta = 0.0001 - J * nu * delta_iter
		if delta < 0:
			print("delta is less than 0")

		return (total_eps - np.sqrt(2 * J * np.log(1 / delta)) * eps_amp - J * eps_amp * (np.exp(eps_amp) - 1)) ** 2

	else:
		# CDP composition
		delta_iter = 0.000001
		delta_amp = nu * delta_iter
		c2 = 2 * np.log(1.25 / delta_amp)
		delta = 0.0001
		rho = J * (eps_amp ** 2) / (2 * c2)

		return (total_eps - (rho + 2 * np.sqrt(rho * np.log(1 / delta)))) ** 2
